oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,486 @@
+
+However, now roughly 5 years into this gradual re-platforming, we've discovered that the completely siloed nature of these MFEs has created its own set of problems. These problems are described below.
+
+Consistency
+===========
+
+Because each MFE is completely independent, this can create inconsistencies from MFE to MFE.
+
+Reusable components (Paragon)
+-----------------------------
+
+MFEs may use different versions of ``@openedx/paragon``, resulting in functional and stylistic differences.
+
+Header/Footer
+-------------
+
+MFEs may use different versions of ``@edx/frontend-component-header`` and ``@edx/frontend-component-footer``, also resulting in functional, stylistic, and content/navigation differences.
+
+Branding
+--------
+
+Brand packages created from ``@edx/brand-openedx`` may be different versions, resulting in any number of subtle visual differences.
+
+Other Dependencies
+------------------
+
+MFEs may have completely different versions of any other dependency. We mitigate some of this by consolidating some important dependencies in ``@edx/frontend-build`` and ``@edx/frontend-platform``, but even those can have different versions from MFE to MFE. For developers, this increases cognitive load and slows velocity because of the need to adjust to the idiosyncracies of each application.
+
+User and Developer Experience
+=============================
+
+Bundle Size
+-----------
+
+Each MFE bundles all of its own dependencies, even if they're the same version as another MFE. This means that as a user navigates between MFEs, they end up downloading the same set of common dependencies over and over again. This can result in megabytes of overhead for the average user navigating between a few MFEs, and slows down the entire experience.
+
+Full Page Refreshes
+-------------------
+
+Each MFE has its own index.html page, and needs to load all its own dependencies from scratch whenever you navigate to it. This means that the browser performs a full page refresh each time a user transfers from MFE to MFE.
+
+Build time
+----------
+
+Each MFE must bundle all its own dependencies at build time, creating significant overhead for the Webpack build process. When building multiple MFEs, this additional, repeated overhead adds up quickly, making builds prohibitively slow.
+
+Dependency Maintenance
+----------------------
+
+Since each MFE has its own complete set of dependencies, the overhead of keeping them all up to date can be overwhelming. Security patches, bug fixes, new features, and breaking changes all add up and create a significant maintenance burden.
+
+Composability
+=============
+
+We have no clean, low-overhead way of composing components from multiple MFEs into a single page, or extending an MFE with additional functionality.
+
+The reality of MFEs is that while each application may represent a cohesive domain or bounded context, sometimes you need content and functionality from more than one domain on the page at the same time.
```

Comment:
Language is hard. :joy:

If I'm remembering my DDD correctly, our MFEs are definitely bounded contexts in that they have their own domain model, complete with modest anti-corruption layers to massage our data into a schema that makes sense within that context. 

In spirit if not always perfectly in practice, our MFEs were _also_ intended to represent a domain or sub-domain of our overall application.  The learning MFE is about learning, course-authoring is about building courses, etc.  Whether those ended up the right granularity is definitely up for debate. 

So the spirit of this sentence is that, practically speaking, a domain or bounded context (an MFE) may need to display stuff from another domain (another MFE) to make a nice experience for users.  MFEs wanna load stuff from each other.  Totally open to a better way of saying this that doesn't feel like it opens up a can of interpretation worms. :joy:

Reply:
```suggestion
The reality of MFEs is that while each application seeks to represent a single cohesive domain or bounded context, sometimes the reality is that content and functionality from more than one domain needs to be on the page at the same time.
```
Removes "you", which I dislike generally, and tries to capture that in _spirit_ we tried to represent one domain/context! I think it could be helpful, as in another comment, that if we seek to assume of the reader a definition of "cohesive domain" or "bounded context", it would be good to link to an authoritative definitional source so there's no confusion.

Reply:
I've tweaked the language here, and added some links out for 'domain' and 'bounded context'